File: IK.py
import numpy
from numpy import *
import math

# Link lengths
a1 = 0.165  # length of link a1 in m
a2 = 0.118  # length of link a2 in m
d1 = 0.166

# End Effector Position
px = float(input("Enter px: "))  # offset of link d3 in m
if -0.1278 <= px <= 0.3310:
    print("Robot is inside the workspace")
else:
    print("Robot is outside the workspace")
    quit()

py = float(input("Enter py: "))  # offset of link d3 in m
if -0.259101506479873 <= py <= 0.329:
    print("Robot is inside the workspace")
else:
    print("Robot is outside the workspace")
    quit()

pz = float(input("Enter pz: "))  # offset of link d3 in m
if 0 <= pz <= 0.166:
    print("Robot is inside the workspace")
else:
    print("Robot is outside the workspace")
    quit()

# Angle theta 2
c2 = (px ** 2 + py ** 2 - a1 ** 2 - a2 ** 2) / (2 * a2 * a1)
s2 = math.sqrt(1 - c2 ** 2)
s2 = numpy.asarray([[s2], [-s2]])

th2 = arccos(c2)
th2 = [th2, -th2]

# Angle theta 1
c1 = ((a1 + a2 * c2) * px + a2 * s2 * py) / ((a2 * s2) ** 2 + (a1 + a2 * c2) ** 2)
s1 = ((a2 * s2 * px) - (a1 + a2 * c2) * py) / ((a2 * s2) ** 2 + (a1 + a2 * c2) ** 2)
th1 = -arctan2(s1, c1)
th1 = [th1, -th1]

# d3
d3 = d1 - pz

print(th1, rad2deg(th1))
print(th2, rad2deg(th2))
print(d3)

File: IK.m
% Simulation of SCARA(RRPR) Robot

syms q1 q2 a1 a2 d3

% q1 q2 q4 joint angles
% d3 link offsets
% a1 a2  link lengths

% Inputs:-

prompt = 'Enter value of q1: ';
q1 = input(prompt)*pi/180;
if ((-99.67*pi/180 <= q1) && (q1 <= 99.67*pi/180))
    disp('robot is inside the workspace')
else 
    disp('robot is outside the workspace!*****')
    return
end
prompt = 'Enter value of q2: ';
q2 = input(prompt)*pi/180;
if ((-112.32*pi/180 <= q2) && (q2 <= 112.32*pi/180))
    disp('robot is inside the workspace')
else 
    disp('robot is outside the workspace!*****')
    return
end
prompt = 'Enter value of d3: ';
d3 = input(prompt);
if ((0 <= d3) && (d3 <= 0.116))
    disp('robot is inside the workspace')
else 
    disp('robot is outside the workspace!*****')
    return
end

% Link lengths are in meter
a1 = 0.165;  
a2 = 0.118;
d1 = 0.166;

% Input matrix using Transformation matrix
A1 = [cos(q1) -sin(q1) 0 a1*cos(q1); sin(q1) cos(q1) 0 a1*sin(q1); 0 0 1  d1; 0 0 0 1];
A2 = [cos(q2) sin(q2) 0 a2*cos(q2); sin(q2) -cos(q2) 0 a2*sin(q2); 0 0 -1 0; 0 0 0 1];
A3 = [1 0 0 0; 0 1 0 0; 0 0 1 d3; 0 0 0 1];

%Robotic Toolbox :- Peter corke
% create a numerical or symbolic DH parameter 
%L   = Link([Th  d    a      alpha offset])
L(1) = Link([0,  d1,  a1,     0,      0]);          % Revolute joint
L(2) = Link([0,  0,   a2,     pi,     0]);          % Revolute joint
L(3) = Link([0,  d3,   0,     0,      1]);          % Prismatic joint
SCARA = SerialLink(L,'name','SCARA');              % seriallink(L) to create the link

% Desired Manipulator matrix by Transformation 
A12 = (A1*A2); 
A13 = (A1)*(A2)*(A3);

P = A13(1:3,4);
disp('End effector position:')
disp(P)

% Inverse kinematics
Q =  SCARA.ikine(A13,'mask',[1 1 1 0 0 0]); 
figure(1)
SCARA.plot(Q,'workspace',[-1 1 -1 1 -1 1])
disp('Joint angles are:')
disp(Q)



